Gossipsub: Partial Message Extension

MarcoPolo opened the pull request:
All the info is in the spec.

MarcoPolo committed:
Partial Messages Extension

Co-authored-by: Csaba Kiraly <csaba.kiraly@gmail.com>

cskiraly commented:
@MarcoPolo proposed improvements to the partial message spec here 
https://github.com/cskiraly/libp2p-specs/tree/csaba/partial-message

In my version I had both partialIHAVE and partialIWANT.

I think I see your point in focusing on the IWANT part, as it might allow us to delegate all the logic to the upper layer, similar to how you can do partial requests in other protocols where you need pieces only (e.g. HTTP Accept-Ranges for resume, and similar use cases). We still need the other node to send the IHAVE (the normal, full IHAVE, and we still would need the logic in our node not to fetch the whole message, so either a hook before IWANT or some call to say "for this I already have something, please send a PartialIWANT with this extra metadata".

If instead we also define the Partial IHAVE, we can do more. In the FullDAS implementation I'm doing these partial IHAVEs to some extent already, because I'm segmenting the column to cell-level messages. We can also give a bitmap encoding to these, as it was proposed there.

Regarding keeping the metadata opaque, I think it is an interesting idea, but it seems to me it requires a more complex API. At the end of the day we do need a protocol layer that is aware of the underlying segment structure. We can try to have this layering of the structure aware and structure unaware part, but I feel this will lead to inefficiencies at some point.

raulk commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,189 @@
+application generates this and consumes this.
+
+
+Nodes SHOULD assume a `partialIWANT `implies a `IDONTWANT `for the full message.
```
```suggestion
Nodes SHOULD assume a `partialIWANT` implies a `IDONTWANT `for the full message.
```

raulk commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,189 @@
+
+### PartialIWANT
+
+Partial IWants signal to a receiver that the sending peer only wants a part of
```
```suggestion
`PartialIWANT`s signal to a receiver that the sending peer only wants a part of
```

raulk commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,189 @@
+segments, unlocking the use of partialIWANT in more contexts.
+
+In the context of partial messages, it is more useful than IHAVE as it tells the
+peer the group ID. In contrast, a IHAVE only includes a message ID that is
```
```suggestion
peer the group ID. In contrast, an IHAVE only includes a message ID that is
```

raulk commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,189 @@
+
+## Motivation
+
+The main motivation for this extension is optimizing Ethereum's Data
+Availability Sampling (DAS) protocol. In Ethereum's upcoming fork, Fusaka,
```
Technically this doesn't benefit the Sampling part of things. Maybe just generalize this to Ethereum's DA protocols.

raulk commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,189 @@
+
+The main motivation for this extension is optimizing Ethereum's Data
+Availability Sampling (DAS) protocol. In Ethereum's upcoming fork, Fusaka,
+custodied data is laid out in a matrix, where the rows represent user data
```
```suggestion
custodied data is laid out in a matrix per block, where the rows represent user data
```

raulk commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,189 @@
+The main motivation for this extension is optimizing Ethereum's Data
+Availability Sampling (DAS) protocol. In Ethereum's upcoming fork, Fusaka,
+custodied data is laid out in a matrix, where the rows represent user data
+(called blobs), and the columns represent a slice across all blobs (each blob
```
```suggestion
(called blobs), and the columns represent a slice across all blobs included in the block (each blob
```

raulk commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,189 @@
+  optional PartialIHAVE ihave = 4;
+}
+
+message PartialMessage {
+  optional bytes topicID = 1;
+  optional bytes data = 2;
+}
```
Would the partial message benefit some contextual metadata to indicate which parts are actually being delivered? Or do we expect the data to encode this (e.g. with null elements inside a list, if that's the shape)? Or does the `PartialMessage` refer to a single part?

Also: this schema only allows for one inflight partial request per peer per topic. I guess that's not a problem and we (Ethereum) are fine with that limitation?

raulk commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,189 @@
+application defined. Here are a list of operations an application is expected to
+provide to Gossipsub to enable partial message delivery.
+
+1. Split a full message into a group of partial messages.
```
We may not have a **full** message, but still benefit from splitting what we have into pieces that can be trasmitted as parts right? (e.g. the common case for PeerDAS columns)

Also, is the term "partial message" is being used ambiguously in this spec? Here it refers to parts, but below it might refer to a collection of parts in response to a PARTIALIWANT?

raulk commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,189 @@
+provide to Gossipsub to enable partial message delivery.
+
+1. Split a full message into a group of partial messages.
+2. Given a group of partial messages, encode a request for the missing parts. (Used by `PartialIWANT`)
```
This will require some higher level contextual information also, right? In our case, in addition to the actual partial messages generated in (1), we need to know the total number of parts and the positions our local parts occupy within the series, so we can generate the bitmap.

jxs commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,191 @@
+
+Partial Messages Extensions allow users to transmit only a small part of a
+message rather than a full message. This is especially useful in cases where
+there is a large messages and a peer is missing only a small part of the
```
```suggestion
there is a large message and a peer is missing only a small part of the
```

MarcoPolo committed:
Remove PartialIDONTWANT

It's purpose is subsumed by PartialIHAVE

sukunrt reviewed (commented):
I want to understand more about how this will interact with the normal(non
partial) message flow on the topic. 

As I understand from:
https://github.com/ethereum/consensus-specs/pull/4558/files we want to disable
normal gossipsub for peers that support partial message extension and rely
completely on partial messages to reconstruct the columns. However we want to
keep the old (non partial) system for peers that don't support partial messages.

If this is the case, implementation wise would it be better to make this a
topic level property where only peers with partial message extension can
participate on a partial message specific topic.

MarcoPolo committed:
gossipsub: add explicit request to use partial messages for a given topic

MarcoPolo commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,191 @@
+belonging to the missing blob. Reducing the network resource usage
+significantly. As an example, if there are 32 blob cells in a column and the
+node has all but one cell, this would result in a transfer of 2KiB rather than
+64KiB per column. and since nodes custody at least 8 columns, the total savings
+per slot is around 500KiB, or 4 Megabits per slot.
```
500KiB*8=4mb. Bits since we are talking about bandwidth, but we can just remove that

MarcoPolo commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,191 @@
+4. `.MissingParts() -> Result<metadata: bytes, Error>` (For `PartialIWANT`)
+5. `.AvailableParts() -> Result<metadata: bytes, Error>` (For `PartialIHAVE`)
+
+Gossipsub in turn provides a `.PublishPartial(PartialMessage)` method.
+
```
the implicit `to` is the peers in a topic mesh, although having a specific to is also useful (the Go implementation supports both implicit of all topic peers and explicit peers)

MarcoPolo replied:
PublishPartial is idempotent. If a peer has sent us a partial iwant, this publish call will respond to it with the relevant parts.

If you call PublishPartial twice with no changes in between, the second time does nothing.

MarcoPolo commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,191 @@
+
+1. `.GroupID() -> GroupID: bytes`
+2. `.PartialMessageBytesFromMetadata(metadata: bytes) -> Result<(EncodedPartialMessage: bytes, metadata: bytes), Error>` (When responding to a `PartialIWANT` or eagerly pushing a partial message)
+  a. The returned metadata represents the still missing parts. For example, if a
+     peer is only able to fulfill a part of the the request, the returned
+     metadata represents the parts it couldn't fulfill.
```
Suggestion for text to make this clearer?

MarcoPolo replied:
I rephrased this, is it clearer?

MarcoPolo commented:
> If this is the case, implementation wise would it be better to make this a topic level property where only peers with partial message extension can participate on a partial message specific topic.

Currently, in the Go implementation, applications themselves decide if a topic should support partial messages by setting a topic option.

Thinking about this again, maybe it would be better to have an explicit flag in the `SubOpts` message that requests the receiver to use partial messages.

sukunrt commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,178 @@
+supports the following operations:
+
+1. `.GroupID() -> GroupID: bytes`
+2. `.PartialMessageBytesFromMetadata(metadata: bytes) -> Result<(EncodedPartialMessage: bytes, metadata: bytes), Error>` (When responding to a `PartialIWANT` or eagerly pushing a partial message)
```
NIT: I prefer `repr` over `metadata` as it is the IWant byte representation, not some meta info about it.

sukunrt replied:
NIT: 
```suggestion
2. `.PartialMessageBytesFromMetadata(IWantMetadata: bytes) -> Result<(EncodedPartialMessage: bytes, newIWantMetadata: bytes), Error>` (When responding to a `PartialIWANT` or eagerly pushing a partial message)
```

sukunrt commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,178 @@
+
+## Application Interface
+
+Message contents are application defined. Thus splitting a message must be
+application defined. Applications should provide a Partial Message type that
+supports the following operations:
```
I'd move the non normative disclaimer here.
 
For the rest of the text, I see two implementation strategies: 
1. Discard all gossipsub infra except sending and receiving messages, and topic mesh handling, and do all the partial message stuff in the application. So all you need from the application is `PublishTo(msg, to)` and `ReceiveRPC(from, rpc)`

2. The one that go-libp2p does where the application periodically does a PublishMessage(msg) providing its view of a particular partial message(topicID, groupID) to the gossipsub layer. The gs layer keeps track of the want and have lists per peer as opaque metadata and updates it based on the methods provided below. 

You may rephrase the text above or use something similar. When reading the document I assumed we'd do 1, and so found this section confusing, but reviewing https://github.com/libp2p/go-libp2p-pubsub/pull/631/files#diff-d45a19dbc01044648070c29c0fc7b2b63edd3deaccf85b797314d5ef991f4db6R48-R51 I can see we do 2, which clears up doubts on my initial readings, like https://github.com/libp2p/specs/pull/685/files#r2379080632

MarcoPolo replied:
We mostly do 2. We try to do as much as we can in the gossipsub layer without making too many assumptions.

MarcoPolo committed:
Simplify the fields and remove message nesting.

There is no longer a separation between iwant/ihave

MarcoPolo commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,178 @@
+
+- Do we want to add a TTL to PartialIWANTs? This would allow us to cancel them after some time.
+- Should we rename the metadata bytes to iwant and ihave?
+- In the bitmap usecase, iwant/ihave are simply inverses of each other. Do we need to send them both?
```
Done in 8e42e179.

I think this simplifies the implementation quite a bit as well.

MarcoPolo committed:
relax the requirement on the SubOpts partial field.

MarcoPolo committed:
Update code Gossipsub partial extension codepoint

MarcoPolo committed:
mark off todos

MarcoPolo committed:
minor formatting

MarcoPolo committed:
Add supportsPartial field to subopts

Let's a peer signal support for partial messages on a per-topic basis.

This makes future rollouts of partial messages on new topics much easier
as nodes can explicitly signal their support of partial messages per
topic.

sukunrt commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -104,20 +104,53 @@ on signaling bandwidth tradeoff considerations.
+topic by setting the `supportsPartial` field in `SubOpts` to true. A node can
+support sending partial messages without wanting to receive them.
+
+If a node requests partial messages, it MUST support partial messages.
```
```suggestion
If a node requests partial messages, it MUST support sending partial messages.
```

sukunrt reviewed (commented):
Can we add a section for safe rollout of partial messages on existing topics? This would be just an implementation/rollout guideline.

MarcoPolo commented:
We need to also add a section for interacting with non-mesh peers.

MarcoPolo committed:
partial-messages: expand application interface section

MarcoPolo committed:
partial-messages: add section on upgrading topics to partial messages

jxs commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,227 @@
+| --------------- | ------------- | ------ | --------------- |
+| 1A              | Working Draft | Active | r0, 2025-06-23  |
+
+Authors: [@marcopolo, @cskiraly]
```
```suggestion
Authors: [@marcopolo], [@cskiraly]
```

jxs commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,227 @@
+
+Authors: [@marcopolo, @cskiraly]
+
+Interest Group: TODO
```
```suggestion
Interest Group: [@jxs], [@dknopik]
```

jxs commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,227 @@
+
+[@marcopolo]: https://github.com/marcopolo
+[@cskiraly]: https://github.com/cskiraly
+
```
```suggestion
[@jxs]: https://github.com/jxs
[@dknopik]: https://github.com/dknopik
```

MarcoPolo committed:
fix outdated description

MarcoPolo committed:
add section about including fannout/gossip peers...

to the list of peers to publish partial messages to.

pawanjay176 commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,224 @@
+The application decides if it should act on the message by either requesting
+parts or forwarding the message. Both are done with `.PublishPartial`.
+
+Gossipsub MUST forward all messages to the application, not just messages from
```
is this not just from mesh peers?

MarcoPolo replied:
yes, thank you.

MarcoPolo committed:
typo

MarcoPolo committed:
Add section on Partial Message Gossip

MarcoPolo committed:
add .EagerPartialMessageBytes() to application interface

MarcoPolo committed:
Add implementation recommendations

MarcoPolo committed:
editorial

MarcoPolo commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,178 @@
+on signaling bandwidth tradeoff considerations.
+
+Receivers MUST treat a `PartialIHAVE` as a signal that the peer does not want
+the indicated part.
```
I think this is covered now.

MarcoPolo committed:
typo

MarcoPolo committed:
formatting

MarcoPolo commented:
@sukunrt, @jxs could I trouble you all for a review? I think this is ready for merging.

MarcoPolo committed:
Add more implementation recommendations

jxs commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,297 @@
+A later `partsMetadata` replaces a prior one.
+
+`partsMetadata` can be used during heartbeat gossip to inform non-mesh topic
+peers about parts this node has.
```
```suggestion
During heartbeat gossip, partsMetadata can be used to inform a random subset of non-mesh topic peers about the parts held by this node, similar to full message IHAVE gossip.
```

jxs commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,297 @@
+### Fanout and Gossip messages
+
+Fanout and Gossip messages by definition come from non-mesh peers. Partial
+messages, without eager data, require an exchange of bitmaps before parts are
```
```suggestion
messages without eager data, require an exchange of bitmaps before parts are
```

jxs commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,297 @@
+
+## Creating a topic to only use partial messages
+
+There is currently no mechanism to requre that a topic only be used for partial
```
```suggestion
There is currently no mechanism to require that a topic only be used for partial
```

jxs commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,297 @@
+
+### Supporting both full and partial messages for a topic
+
+Partial messages use the same mesh as "full" messages. Supporting both is
```
should it be
```suggestion
Partial messages use the same mesh as full messages. Supporting both is
```
instead?
as we already define above the concept of full messages

jxs commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,297 @@
+   1. The method should return an encoded partial message of eager data that
+      should be sent, along with the partsMetadata a peer would have after
+      receiving the encoded partial message.
+4. `.PartsMetadata() -> bytes`: The parts this partial message has.
```
this API is now different in the `rust-libp2p`  [implementation](https://github.com/jxs/rust-libp2p/blob/gossipsub-partial-messages/protocols/gossipsub/src/extensions/partial_messages.rs#L72-L93).

MarcoPolo replied:
I'll rework this section from our latest learnings.

jxs reviewed (approved):
left some suggestions Marco but overall LGTM

sukunrt commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,297 @@
+peer did not send a value.
+
+Upon receiving a `partsMetadata` a node SHOULD respond with only parts the peer
+wants.
```
```suggestion
Upon receiving a `partsMetadata` a node SHOULD respond with only parts the peer
doesn't have.
```

sukunrt commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,297 @@
+partial message request to a peer. It uses its `requestsPartial` setting to know
+whether to send the peer a full message or a partial message.
+
+If a peer supports partial messages on a topic but did not request them, a node
```
```suggestion
If a peer supports sending partial messages on a topic but did not request them, a node
```

sukunrt commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,297 @@
+## Partial Message Gossip
+
+Partial Messages can replace Gossipsub's IHAVE/IWANT with a message that
+provides more context (via the Group ID) and allows for partial responses.
```
```suggestion
Partial Messages SHOULD replace Gossipsub's IHAVE/IWANT with a message that
provides more context (via the Group ID) and allows for partial responses.
```

sukunrt commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,297 @@
+of the its previously sent `partsMetadata `and the remote peer's
+`partsMetadata`. The node does not need to send this peer its updated
+`partsMetadata` if its `partsMetadata` has not changed since the previously sent
+`partsMetadata`.
+
+On the other side, when a node responds to a peer's request for parts, it should
+update the peer's `partsMetadata` with the union of its `partsMetadata` and the
+peers `partsMetadata`.
+
```
I think this isn't true anymore. This spec doesn't enforce or even suggest sending all the parts that the node has and the peer doesn't have to the peer. So the union op is incorrect.

MarcoPolo replied:
you're right. I'll remove this section.

vladopajic commented on pubsub/gossipsub/extensions/extensions.proto:
```diff
@@ -23,17 +23,39 @@ message RPC {
+}
+
+message PartialMessagesExtension {
+  optional bytes topicID = 1;
+  optional bytes groupID = 2;
```
why are these fields optional?

vladopajic replied:
since `groupID` identifies a full logical message, it only makes sense for it to always be specified. feels to me like this should be a `MUST`.

MarcoPolo replied:
The protocol requires these, but in protobuf terms these are optional. "Semantics for required fields should be implemented at the application layer instead." from https://protobuf.dev/programming-guides/proto2/#required-deprecated

vladopajic replied:
got it

sukunrt commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,297 @@
+If a node requests partial messages and is connected partial message capable
+peers, it MAY prefer to delay reacting to a peer's `IHAVE` message in order to
+give the opportunity for a partial message request to finish and provide the
+missing message more efficiently.
```
```suggestion
If a node is reconstructing a message with partial message extension, it MAY prefer to delay reacting to a peer's `IHAVE` message in order to
give the opportunity for a partial message request to finish and get the message more efficiently.
```

sukunrt commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,297 @@
+
+When Gossipsub receives a partial message it MUST forward it to the application.
+The application decides if it should act on the message by either requesting
+parts or forwarding the message. Both are done with `.PublishPartial`.
```
Should this be outside of the implementation guidelines heading?

MarcoPolo replied:
This is up to the implementation, right?

sukunrt replied:
this specific part isn't. If the partsMetadata is opaque, the implementation MUST necessarily invoke the application to see what to do with the incoming partial message.

sukunrt commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,297 @@
+parts or forwarding the message. Both are done with `.PublishPartial`.
+
+Gossipsub MUST forward all messages to the application, not just messages from
+mesh peers.
```
Same comment on this one. Should this be outside implementation guidelines?

sukunrt commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,297 @@
+mesh peers.
+
+### Fanout and Gossip messages
+
```
Same comment here. Move it outside implementation recommendations.

MarcoPolo replied:
I'm not sure, fanout and gossip and mesh peers are identical from the protocol's perspective. How this is presented is up to the implementation. What do you think should go outside the implementation guidelines section?

sukunrt replied:
on this point + https://github.com/libp2p/specs/pull/685#discussion_r2789068407 combined:

I would prefer calling out that a partial message irrespective of the sender (gossip, mesh, fanout) MUST be pushed to the application. There's nothing gossipsub layer can do here.

sukunrt commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,297 @@
+transferred. In order for fanout and gossip messages to be useful, the Gossipsub
+implementation SHOULD include the peers that sent them in the list of peers to
+publish partial messages to. This allows the application to simply call
+`PublishPartial` to respond to both mesh and non mesh peers.
```
NIT
```suggestion
Fanout and Gossip messages by definition come from non-mesh peers. After receiving partial messages from Fanout and Gossip peers, the Gossipsub
implementation SHOULD include these non mesh peers that sent them in the list of peers to
publish partial messages to. This allows the application to simply call
`PublishPartial` to respond to both mesh and non mesh peers.
```

sukunrt commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,297 @@
+there is a large message and a peer is missing only a small part of the
+message.
+
+## Terms and Definitions
+
+**Full Message**: A Gossipsub Message.
+
+**Message Part**: The smallest verifiable part of a message.
+
+**Partial Message**: A group of one or more message parts.
+
+**Group ID**: An identifier to some Full Message. This must not depend on
+knowing the full message, so it can not simply be a hash of the full message.
```
Can you add Eager Data here?

sukunrt reviewed (approved):
Mostly nits and rephrasings.

ackintosh commented on pubsub/gossipsub/partial-messages.md:
```diff
@@ -0,0 +1,297 @@
+If a node requests partial messages, it MUST support sending partial messages.
+
+A node uses a peer's `supportsSendingPartial` setting to know if it can send a
+partial message request to a peer. It uses its `requestsPartial` setting to know
```
What does "partial message request" mean? IIUC, it means sending `partsMetadata` to request missing parts from a peer. If so, I'd suggest making this explicit: 


```suggestion
A node uses a peer's `supportsSendingPartial` setting to know if it can send `partsMetadata` to a peer. It uses its `requestsPartial` setting to know
```

MarcoPolo replied:
Thanks, updated in the latest commit.

MarcoPolo committed:
Editorial changes from review

MarcoPolo committed:
Add Application-Library Interface section

MarcoPolo committed:
Refine implementation recommendations

MarcoPolo committed:
Add Minimizing unnecessary messages section

MarcoPolo commented:
This should be good to go. I'll merge this first thing tomorrow unless there's an objection.

Recent changes have been editorial, while the protocol itself has been fairly stable.

MarcoPolo committed:
Add co-authors
